zebra-rpc/src/methods.rs

```diff
@@ -526,6 +542,76 @@ where
         }
         .boxed()
     }
+
+    fn get_address_utxos(&self, addresses: Vec<String>) -> BoxFuture<Result<Vec<AddressUtxos>>> {
+        let mut state = self.state.clone();
+        let mut response_utxos = vec![];
+
+        async move {
+            // TODO: make this a function after #4062 is merged
+            let valid_addresses: Result<Vec<Address>> = addresses
+                .iter()
+                .map(|address| {
+                    address.parse().map_err(|_| {
+                        Error::invalid_params(format!("Provided address is not valid: {}", address))
+                    })
+                })
+                .collect();
+
+            // get utxos data for addresses
+            let request = zebra_state::ReadRequest::UtxosByAddresses(valid_addresses?);
+            let response = state
+                .ready()
+                .and_then(|service| service.call(request))
+                .await
+                .map_err(|error| Error {
+                    code: ErrorCode::ServerError(0),
+                    message: error.to_string(),
+                    data: None,
+                })?;
+            let utxos = match response {
+                zebra_state::ReadResponse::Utxos(utxos) => utxos,
+                _ => unreachable!("unmatched response to a UtxosByAddresses request"),
+            };
+
+            // for each utxo found, get transaction
+            for utxo in utxos {
+                // given hash, get transaction
+                let hash = utxo.output_location.hash;
```

Comment:
How do i go from `TransactionLocation` to transaction `Hash` or `Transaction` ? Not sure if this is currently possible.

Reply:
I'm still working on this query, but here's my plan:
* combine the non-finalized `TransparentTransfers::created_utxos` and finalized `ZebraDb::address_utxos` for all the addresses
* remove the `TransparentTransfers::spent_utxos`

To do that, I need to change `ZebraDb::address_utxos` so it also returns transaction hashes.

These queries depend on these methods in these PRs:

`TransparentTransfers::created_utxos` (PR #4022)
https://github.com/ZcashFoundation/zebra/blob/0031529669b5eded8a4843e252bfde1bcc69041e/zebra-state/src/service/non_finalized_state/chain/index.rs#L243-L255

`TransparentTransfers::spent_utxos` (PR #4022)
https://github.com/ZcashFoundation/zebra/blob/0031529669b5eded8a4843e252bfde1bcc69041e/zebra-state/src/service/non_finalized_state/chain/index.rs#L268-L271

`ZebraDb::address_utxos` (PR #4038)
https://github.com/ZcashFoundation/zebra/blob/5a4ba1ad26ea4150bb0541089147579a06a9e6ca/zebra-state/src/service/finalized_state/zebra_db/transparent.rs#L101-L107